### PR TITLE
Fix generating the `Origin` header value for websocket handshake request

### PR DESCRIPTION
**Pull request by amizurov**
Motivation:

We have the old erroneous behavior of generating the `Origin| Sec-WebSocket-Origin` for client websocket handshake request (https://github.com/netty/netty/issues/9673). In Netty5 this fixed and auto-generation has been deleted at all, only if the client passed the `Origin` header via custom headers. The same we can do for Netty4 but it could potentially break some clients (unlikely), or introduce an additional parameter to disable or enable this behavior.

Modification:

Introduce new `generateOriginHeader` parameter in client config and generate the `Origin|Sec-WebSocket-Origin` header value only if it enabled. Add additional check for webSocketURI if it contains host or passed through `customHeaders` to prevent NPE in `newHandshakeRequest()`.

Result:

Fixes https://github.com/netty/netty/issues/9673 https://github.com/netty/netty/issues/12933

### CODE REVIEW

**Comment by amizurov** [comment]
@normanmaurer , @chrisvest @vietj , @slandelle Could you please take a look.

**Comment by chrisvest** [comment]
One fixes the NPE by turning origin header generation off?

**Comment by ursaj** [comment]
looks good

**Comment by amizurov** [comment]
> One fixes the NPE by turning origin header generation off?

Added a comment to the issue https://github.com/netty/netty/issues/12933 how they are related.

**Review by chrisvest — APPROVED** [review]

**Comment by amizurov** [comment]
@normanmaurer @vietj @slandelle Could you please review this changes, i would like to know your thoughts.

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java [hunk 1]
```diff
@@ -192,6 +195,62 @@ public void testSetOriginFromCustomHeaders() {
...
+        final WebSocketClientHandshaker handshaker = newHandshaker(uri, null, headers, false, true);
+        final ChannelFuture handshakeFuture = handshaker.handshake(channel);
+
+        assertFalse(handshakeFuture.isSuccess());
+        Throwable exception = assertThrows(IllegalArgumentException.class, new Executable() {
+            @Override
+            public void execute() throws Throwable {
+                throw handshakeFuture.cause();
+            }
+        });
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java:246]: This is an odd usage of assertThrows:

```suggestion
        assertInstanceOf(IllegalArgumentException.class,  handshakeFuture.cause());
```

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java [hunk 2]
```diff
@@ -192,6 +195,62 @@ public void testSetOriginFromCustomHeaders() {
...
+        final WebSocketClientHandshaker handshaker = newHandshaker(uri, null, null, false, true);
+        final ChannelFuture handshakeFuture = handshaker.handshake(channel);
+
+        assertFalse(handshakeFuture.isSuccess());
+        Throwable exception = assertThrows(IllegalArgumentException.class, new Executable() {
+            @Override
+            public void execute() throws Throwable {
+                throw handshakeFuture.cause();
+            }
+        });
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java:224]: This is an odd usage of assertThrows:

```suggestion
        assertInstanceOf(IllegalArgumentException.class,  handshakeFuture.cause());
```
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Just some comments related to the test-code. PTAL

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java (hunk 1 above)
Comment (amizurov) [reply codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java]: Done

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java (hunk 2 above)
Comment (amizurov) [reply codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshakerTest.java]: Done

**Comment by normanmaurer** [comment]
@amizurov thanks a lot!

**Review by slandelle — COMMENTED** [review]
LGTM. I'm fine with fixing a broken behavior at the cost of a breaking change.

**Pull request merged**